www/script.js

```diff
@@ -0,0 +1,186 @@
+mapping["cyclone"] = "c5";
+mapping["venom"] = "c6";
+mapping["drought"] = "c7";
+mapping["volcano"] = "c8";
+mapping["blight"] = "c9";
+mapping["hailstorm"] = "c10";
+mapping["arsenic"] = "c11";
+mapping["heatwave"] = "c12";
+mapping["cyanide"] = "c13";
+mapping["acid"] = "c14";
+mapping["invasion"] = "c15";
+mapping["eruption"] = "c16";
+mapping["headcrash"] = "c17";
+mapping["asteroid"] = "c18";
+mapping["avalanche"] = "c19";
+mapping["blackout"] = "c20";
+mapping["plague"] = "c21";
+mapping["bigbang"] = "c22";
+mapping["wildfire"] = "c23";
+mapping["famine"] = "c24";
+mapping["madcow"] = "c25";
+mapping["surge"] = "c26";
+mapping["meteorstorm"] = "c27";
+mapping["chaos"] = "c28";
+mapping["blizzard"] = "c29";
+mapping["hurricane"] = "c32";
+
+/**
+ * Convert a time like "12:30:00" to a shortened time like "12:30 pm"
+ *
+ * 12:00:00 -> 12 pm
+ * 16:30:00 -> 4:30 pm
+ * 09:15:15 -> 9:15:15 am
+ */
+function shortTime(t) {
+    var ints = t.split(':').map(x => parseInt(x));
+
+    // only show significant parts of the time
+    while (ints[ints.length - 1] === 0) {
+        ints.pop();
+    }
+
+    // midnight special case
+    if (ints.length === 0) {
+        return '12 am';
+    }
+
+    var suffix = 'am';
+    if (ints[0] >= 12 && ints[0] !== 24) {
+        suffix = 'pm';
+    }
+    if (ints[0] === 0) {
+        ints[0] = 12;
+    }
+    if (ints[0] > 12) {
+        ints[0] -= 12;
+    }
+    return ints.join(':') + ' ' + suffix;
+}
+
+function hasOverlap(x, set) {
+    for (e of x) {
+        if (set.has(e)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// HTTP GET that returns a promise
+function get(url) {
+  return new Promise(function(resolve, reject) {
+    var req = new XMLHttpRequest();
+    req.open('GET', url);
+
+    req.onload = function() {
+      if (req.status == 200) {
+        resolve(req.response);
+      }
+      else {
+        reject(Error(req.statusText));
+      }
+    };
+
+    req.onerror = function() {
+      reject(Error("Network Error"));
+    };
+
+    req.send();
+  });
+}
+
+function desktops_in_use() {
+    return get('https://www.ocf.berkeley.edu/api/lab/desktops').then(function(response) {
+        return JSON.parse(response)['public_desktops_in_use'];
+    });
+}
+
+function hours_today() {
+    return get('https://www.ocf.berkeley.edu/api/hours/today').then(function(response) {
+        return JSON.parse(response);
+    });
+}
+
+function nightMode() {
+    document.body.classList.add('nightmode');
+}
+
+function updateMap(desktops_in_use) {
+    var id_in_use = new Set();
+    for (const desktop_name of desktops_in_use) {
+        id_in_use.add(mapping[desktop_name]);
+    }
+    // Temporary fix to always show eruption and destruction as being in use.
+    // TODO: there should be an API for public_dekstops_free, which we
+    //       should be using instead. That better represents the purpose
+    //       of this utility anyways.
+    id_in_use.add(mapping['eruption']);
```

Comment:
I believe this API endpoint does exist now? public_desktops_in_use?

Reply:
In my mind, the difference between "free" and "in-use" is that "free" would exclude eruption (and computers that are off) server-side, whereas "in-use" wouldn't necessarily give us that information. So they're not inverses of eachother.